apis/kernel/ipc/src/lib.rs

```diff
@@ -0,0 +1,349 @@
+    /// calling `Ipc::notify_service` with the current service's service ID.
+    pub fn register_service_listener<F: Fn(IpcCallData)>(
+        pkg_name: &[u8],
+        listener: &'static IpcListener<F>,
+    ) -> Result<(), ErrorCode> {
+        let service_id = Self::discover(pkg_name)?;
+        Self::subscribe_ipc::<C, _, DRIVER_NUM>(service_id, listener)
+    }
+
+    /// Register a client IPC callback
+    ///
+    /// This function is called by the IPC client to register a listener
+    /// (callback) for a given IPC service, identified by its service ID. A
+    /// single callback can be registered per service on each client. The
+    /// corresponding service can trigger this callback using
+    /// `Ipc::notify_slicent`, returning control to the user.
+    pub fn register_client_listener<F: Fn(IpcCallData)>(
+        service_id: u32,
+        listener: &'static IpcListener<F>,
+    ) -> Result<(), ErrorCode> {
+        Self::subscribe_ipc::<C, _, DRIVER_NUM>(service_id, listener)
+    }
+
+    /// Notify an IPC service to run
+    ///
+    /// This function is called by the IPC client to trigger an IPC service
+    /// to run. The service ID passed to this function is the same one that
+    /// `Ipc::discover` returns.
+    pub fn notify_service(service_id: u32) -> Result<(), ErrorCode> {
+        S::command(DRIVER_NUM, command::SERVICE_NOTIFY, service_id, 0).to_result()
+    }
+
+    /// Notify a client IPC callback to run
+    ///
+    /// This function is called by the IPC service, generally as part of its
+    /// listener, to trigger an IPC client callback to run. The client ID
+    /// passed to this function is the same one that is presented in the
+    /// `caller_id` field of the `IpcCallData` when the IPC service
+    /// listener executes.
+    pub fn notify_client(client_id: u32) -> Result<(), ErrorCode> {
+        S::command(DRIVER_NUM, command::CLIENT_NOTIFY, client_id, 0).to_result()
+    }
+
+    /// Share a read/write buffer with an IPC service
+    ///
+    /// The client can call this function with a static mutable buffer in order
+    /// to share it via a read/write allow with an IPC service. Since the
+    /// buffer must be mutable and have static lifetime, the best way to
+    /// create a reference to it is via a TakeCell (see the `takecell` crate).
+    pub fn share(service_id: u32, buffer: &'static mut [u8]) -> Result<(), ErrorCode> {
+        Self::allow_rw_ipc::<C, DRIVER_NUM>(service_id, buffer)
+    }
+}
+
+/// A wrapper around a function to be registered and called when an IPC notify
+/// is received.
+///
+/// The IPC API for registering listeners accepts static references to
+/// instances of this struct, so in general the IPC function this struct
+/// wraps should be an actual function instead of a short-lived closure.
+pub struct IpcListener<F: Fn(IpcCallData)>(pub F);
+
+impl<F: Fn(IpcCallData)> Upcall<subscribe::AnyId> for IpcListener<F> {
```

Comment:
Any reason the callback Fn argument is a struct rather than two arguments? (e.g.: `Fn(caller_id: u32, buffer: &Option<&mut [u8]>)`

Reply:
I'd only put it in a struct to parallel the structure of some of the other drivers that register listeners, but in retrospect I agree having it just as `Fn(caller_id: u32, buffer: &Option<&mut [u8]>)` makes the most sense.